### ISSUE TITLE
onPaymentError() always called after payment amount is successfully deducted in Android app release build variant

### ISSUE THREAD

**Issue by FireAndIce** [razorpay/razorpay-android-sample-app#274]
## Description

In the release build variant, the onPaymentError() is always being called, even after the payment amount is successfully deducted.

On investigating the arguments in the onPaymentError(), the following response is being received from Razorpay.

Response in onPaymentError():

```
{
  "code": "BAD_REQUEST_ERROR",
  "description": "Your payment has been declined as the order is invalid. To pay successfully try initiating the payment again from merchant checkout.",
  "source": "business",
  "step": "Payment_initiation",
  "reason": "input_validation_failed",
  "metadata": {
    "order_id": "order_IPoe3Oyg1mTW"
  }
}
```

I have also checked the checkout options. They are as below:
```
{
  "name": "AAAA ZZZZ",
  "description": "Bill Payment 404",
  "order_id": "order_IPoe3Oyg1mTW",
  "allow_rotation": true,
  "currency": "INR",
  "amount": 100,
  "prefill": {
    "email": "hhhh@gmail.com",
    "contact": "901xxxxxxx"
  },
  "method": {
    "netbanking": false,
    "card": false,
    "wallet": false,
    "upi": true
  },
  "readonly": {
    "email": 1,
    "contact": 1
  },
  "notes": {
    "unitName": "404",
    "unitOwnerName": "VxxxKsxxx"
  }
}
```

What could be causing the issue?

## Razorpay SDK Version :
def razorpayVersion = '1.6.15'
implementation 'com.razorpay:checkout:' + razorpayVersion

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1.
2.

## Code example, screenshot, or link to a repository:
![Screenshot from 2021-11-25 17-07-01](https://user-images.githubusercontent.com/1366985/143531491-feaf1a05-c39d-4fbf-904a-189270835f6c.png)
![Screenshot from 2021-11-25 17-07-26](https://user-images.githubusercontent.com/1366985/143531497-ebccd1bf-4dea-4250-b93d-63094ceca783.png)

**Multiple order ids have the same receipt no. Could that be the issue?**

**Comment by ramya12345** [comment]
As per the Important Notes  mentioned in docs here>>>>>  https://razorpay.com/docs/payment-gateway/android-integration/standard/
![razorpayissue](https://user-images.githubusercontent.com/18005152/143905183-fbaa966d-ebd4-4799-b5e7-2b2ee294b656.png)

   According to PCI regulations, payment processing is not allowed on TLS v1. Hence, if the device does not have TLS v1.1 or TLS v1.2, the SDK will throw an error in the onPaymentError method.

**Comment by FireAndIce** [comment]
Starting API Level 20+, TLS is enabled by default. So that should not be the issue. https://developer.android.com/reference/javax/net/ssl/SSLSocket

**Comment by FireAndIce** [comment]
The logs of an entire transaction done on the device can be found here. https://docs.google.com/document/d/1TK1MiKcrU4FzSk_utJ_eWZnv8acOtScmwk94lLw6J70/edit?usp=sharing

In the crash logs I have highlighted few items that could be of relevance. Please look at that too. 

Configuration used for the transaction:
1. Android hardware device
2. Razorpay Live Mode
3. GPay used for the transaction
4. Android App in release mode

**Comment by ArunM03** [comment]
Again the same error appearing, How to solve this ? I'm using 1.6.21. Money deducted successfully but onPaymentError Method called after payment.
I am using razorpay for the past one year to most of my clients. I checked in their apps as well the same error showing, Please fix this asap.

**Comment by kmrinmoy07** [comment]
I am also facing this issue, earlier it was not happening

**Comment by himanshu-nain** [comment]
**UPDATE (Aug 23):**  This issue hasn't been resolved as of Aug 23. I've reported this issue to razorpay technical support team multiple times but got no concrete support so far. This bug appears with certain set of users who are using UPI payment method at random times and is often followed by sdk reporting INVALID_CHECKOUT_OPTIONS error (magically disappers when user try payment again); therefore hard to debug.

**Comment by vivekshindhe** [comment]
Hey @himanshu-nain can you send me a couple of order_ids where this has happened? You can mail it over at vivek.shindhe@razorpay.com. The public key would be great as well.

**Comment by himanshu-nain** [comment]
> Hey @himanshu-nain can you send me a couple of order_ids where this has happened? You can mail it over at [vivek.shindhe@razorpay.com](mailto:vivek.shindhe@razorpay.com). The public key would be great as well.

Hi @vivekshindhe, I've shared required details with you over the mail.

**Comment by kmrinmoy07** [comment]
I am also facing similar issues, in my custom checkout - I have seen that after the subscription is initiated and the user has successfully made the payment and the subscription is activated, but still in 'PaymentResultWithDataListener', I am getting response in 'onPaymentError' instead of getting response in 'onPaymentSuccess'. but in onPaymentError we used to get response as - '{"error":{"code":"BAD_REQUEST_ERROR","description":"Payment processing cancelled by user","source":"customer","step":"payment_authentication","reason":"payment_cancelled","metadata":{"payment_id":"pay_XXXXXXXXX"}}}'. but you can see that payment id is also generated, which means payment is successful

**Comment by himanshu-nain** [comment]
> > Hey @himanshu-nain can you send me a couple of order_ids where this has happened? You can mail it over at [vivek.shindhe@razorpay.com](mailto:vivek.shindhe@razorpay.com). The public key would be great as well.
> 
> Hi @vivekshindhe, I've shared required details with you over the mail.

Hi @vivekshindhe, any update on this??

**Comment by debz-eight** [comment]
> I am also facing similar issues, in my custom checkout - I have seen that after the subscription is initiated and the user has successfully made the payment and the subscription is activated, but still in 'PaymentResultWithDataListener', I am getting response in 'onPaymentError' instead of getting response in 'onPaymentSuccess'. but in onPaymentError we used to get response as - '{"error":{"code":"BAD_REQUEST_ERROR","description":"Payment processing cancelled by user","source":"customer","step":"payment_authentication","reason":"payment_cancelled","metadata":{"payment_id":"pay_XXXXXXXXX"}}}'. but you can see that payment id is also generated, which means payment is successful

I am getting the exact same error. Once the payment is done for subscription via UPI in test mode using custom checkout, the subscription is showing activated, even backend is getting the webhook of subscription state activated but onPaymentError() is getting trigerred.

**Comment by vivekshindhe** [comment]
@debz-eight which application are you using for the UPI AutoPay payment? The general conclusion is UPI AutoPay payments or UPI Intent payments don't work in test mode. It's always successful regardless of what data the UPI application sends back to the SDK. Can you try it with a live key?

**Comment by debz-eight** [comment]
> @debz-eight which application are you using for the UPI AutoPay payment? The general conclusion is UPI AutoPay payments or UPI Intent payments don't work in test mode. It's always successful regardless of what data the UPI application sends back to the SDK. Can you try it with a live key?

I am using Android Custom SDK to implement.

implementation 'com.razorpay:customui:3.9.20'

Here is a 1 min video of what I am doing and getting:

https://drive.google.com/file/d/1w6jT2kwLJcXCLhGwnVWYu8BV3-K3Aa5e/view?usp=sharing

I'll try in the live mode once.

**Comment by debz-eight** [comment]
@vivekshindhe I am putting some parts of the code which is relevant here, for others to check it. Also thanks for the reply, I was just about to create a new issue.

###What I was facing:

I have seen that after the subscription is initiated and the user has successfully made the payment and the subscription is activated, but still in 'PaymentResultWithDataListener', I am getting response in 'onPaymentError' instead of getting response in 'onPaymentSuccess'. but in onPaymentError we used to get response as -

```
{
  "error": {
    "code": "BAD_REQUEST_ERROR",
    "description": "You may have cancelled the payment or there was a delay in response from the UPI app",
    "source": "customer",
    "step": "payment_authentication",
    "reason": "payment_cancelled",
    "metadata": {
      "payment_id": "pay_O2mkOBGptiWo0s"
    }
  }
}
```

This is the data I am submitting to the Razorpay via object:

```
razorpay = Razorpay(this, BuildConfig.RAZORPAY_TEST_API_KEY_ID)
razorpay?.setWebView(razorPayCallbackData.webView)
val data = JSONObject()
data.put("amount", 7900)
data.put("email", "debz.grad@gmail.com")
data.put("contact", "9474135500")
data.put("currency", "INR")
data.put("subscription_id", razorPayCallbackData.subscriptionId)
data.put("method", "upi") // Method specific fields
data.put("_[flow]", "intent")
data.put("upi_app_package_name", "com.google.android.apps.nbu.paisa.user")
```

After this I am submitting the data:

```
razorpay?.submit(data, object : PaymentResultWithDataListener {
                override fun onPaymentSuccess(razorpayPaymentId: String, data: PaymentData) {
                    dismissBusyDialog()
                    toggleWebViewVisibility(razorPayCallbackData.webView, razorPayCallbackData.appBarLayout, razorPayCallbackData.nestedScrollView, false)

                    showLog(data.data.toString(), "RazorPaySuccess")

                }

                override fun onPaymentError(p0: Int, p1: String?, p2: PaymentData?) {
                    dismissBusyDialog()
                    toggleWebViewVisibility(razorPayCallbackData.webView, razorPayCallbackData.appBarLayout, razorPayCallbackData.nestedScrollView, false)

                    showLog(p2.toString(), "RazorPayFailure")
                    showLog(p1, "RazorPayFailure")
                    showLog(p0.toString(), "RazorPayFailure")
                }
            })
```

The toggleWebview visibility function is used to make the exisitng views gone and make webview visible/gone depending on the condition. The dismissBusyDialog() is used to control out own custom loader

After this I was getting the error that I mentioned in ###What I was facing.
